numpy/core/src/multiarray/dtypemeta.c

```diff
@@ -0,0 +1,237 @@
+                "use `np.dtype` instead.");
+        return NULL;
+    }
+    Py_INCREF(self->singleton);
+    return (PyObject *)self->singleton;
+}
+
+/**
+ * This function takes a PyArray_Descr and replaces its base class with
+ * a newly created dtype subclass (DTypeMeta instances).
+ * There are some subtleties that need to be remembered when doing this,
+ * first for the class objects itself it could be either a HeapType or not.
+ * Since we are defining the DType from C, we will not make it a HeapType,
+ * thus making it identical to a typical *static* type (except that we
+ * malloc it). We could do it the other way, but there seems no reason to
+ * do so.
+ *
+ * The DType instances (the actual dtypes or descriptors), are based on
+ * prototypes which are passed in. These should not be garbage collected
+ * and thus Py_TPFLAGS_HAVE_GC is not set. (We could allow this, but than
+ * would have to allocate a new object, since the GC needs information before
+ * the actual struct).
+ *
+ * The above is the reason why we should works exactly like we would for a
+ * static type here.
+ * Otherwise, we blurry the lines between C-defined extension classes
+ * and Python subclasses. e.g. `class MyInt(int): pass` is very different
+ * from our `class Float64(np.dtype): pass`, because the latter should not
+ * be a HeapType and its instances should be exact PyArray_Descr structs.
+ *
+ * @param descr The descriptor that should be wrapped.
+ * @param name The name for the DType, if NULL the type character is used.
+ *
+ * @returns 0 on success, -1 on failure.
+ */
+NPY_NO_EXPORT int
+dtypemeta_wrap_legacy_descriptor(PyArray_Descr *descr)
+{
+    if (Py_TYPE(descr) != &PyArrayDescr_Type) {
+        PyErr_Format(PyExc_RuntimeError,
+                "During creation/wrapping of legacy DType, the original class "
+                "was not PyArrayDescr_Type (it is replaced in this step).");
+        return -1;
+    }
+
+    /*
+     * Note: we have no intention of freeing the memory again since this
+     * behaves identically to static type definition (see comment above).
+     * This is much cleaner for the legacy API, in the new API both static
+     * and heap types are possible.
+     * In particular our own DTypes can be true static declarations.
+     * However, this function remains necessary for legacy user dtypes.
+     */
+    char *tp_name = PyDataMem_NEW(100);
+    if (tp_name == NULL) {
+        PyErr_NoMemory();
+        return -1;
+    }
+    snprintf(tp_name, 100, "numpy.dtype[%s]",
+             descr->typeobj->tp_name);
+
+    PyArray_DTypeMeta *dtype_class = PyDataMem_NEW(sizeof(PyArray_DTypeMeta));
+    if (dtype_class == NULL) {
+        PyDataMem_FREE(tp_name);
+        return -1;
+    }
+    /*
+     * Initialize the struct fields similar to static code:
+     */
+    /* Copy in np.dtype since it shares most things */
+    memcpy(dtype_class, &PyArrayDescr_Type, sizeof(PyArray_DTypeMeta));
```

Comment:
You shouldn't copy from the start of the object. It overrides the PyObject fields, which for PyPy holds the link to the PyPy object. I am pretty sure this is what is causing the segfaults on PyPy. So something like
```
memcpy((char*)dtype_class + sizeof(PyObject), (char*)&PyArrayDescr_Type + sizeof(PyObject), sizeof(PyArray_DTypeMeta) - sizeof(PyObject));
```

I also had this [problem](https://github.com/numpy/numpy/pull/12585/files#diff-f3387afa660b73f8f37f0978193799d9R4442)

Reply:
Ah then a `PyObject_Init()` on the type should be the right thing, lets see (I was not able to reproduce locally, strangely enough, even with the pypy nightly).

Reply:
OK, after a lot of pain (and still not being quite sure where things go wrong, but maybe subtly different things at different times), it must indeed be the incorrect initialization of the `Py(Var)Object` fields. This now seems to work fine, by creating a new static "prototype" which is copied. That ensures that the code indeed looks identical to a static type declaration, instead of approximating it in a way that was not quite good enough for PyPy.